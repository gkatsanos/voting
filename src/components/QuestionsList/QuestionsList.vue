<template>
  <div>
    <h4>Questions</h4>
    <v-layout row wrap>
      <question
        v-for="(question, index) in questions"
        :key="question.url"
        :id="question.url"
        :question="question">
      </question>
    </v-layout>
    <v-layout column align-center>
      <v-btn
          data-qa="load-more-questions"
          large
          color="primary"
          :block="$vuetify.breakpoint.xsOnly"
          v-if="!maxPagesReached"
          @click="$store.dispatch('requestItems')">
        Load More Questions</v-btn>
    </v-layout>
  </div>
</template>

<script>
  import { mapGetters, mapState } from 'vuex';
  import Question from '../Question/Question';

  export default {
    name: 'questionsList',
    components: {
      Question,
    },
    computed: {
      ...mapGetters([
        'questions',
      ]),
      ...mapState([
        'maxPagesReached',
      ]),
    },
    mounted() {
      if (!this.$store.state.hasLoaded) {
        this.$store.dispatch('requestItems');
      }
    },
  };
</script>
