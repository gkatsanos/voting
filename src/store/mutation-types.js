export const PRE_HTTP_REQUEST = 'PRE_HTTP_REQUEST';
export const FETCHING_ITEMS = 'FETCHING_ITEMS';
export const FETCHED_ADS_SUCCESS = 'FETCHED_ADS_SUCCESS';
export const INCREASE_PAGE = 'INCREASE_PAGE';
export const DETAIL_VIEW = 'DETAIL_VIEW';
export const SET_URL = 'SET_URL';
export const BTN_LOADING = 'BTN_LOADING';
export const VOTE = 'VOTE';
export const QUESTION_SAVED = 'QUESTION_SAVED';
export const RESET_QUESTION_FLAG = 'RESET_QUESTION_FLAG';
export const LOAD_MORE_QUESTIONS = 'LOAD_MORE_QUESTIONS';
export const FETCHED_ADS_FAIL = 'FETCHED_ADS_FAIL';
