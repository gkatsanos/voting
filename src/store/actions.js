import api from '../api/api';
import * as types from './mutation-types';

export const requestItems = ({ commit, dispatch }) => {
  dispatch('getUrl');
  commit(types.PRE_HTTP_REQUEST);
};

export const getUrl = ({ commit, dispatch }) => {
  api.getUrl().then((data) => {
    commit(types.SET_URL, data);
    dispatch('fetchItems');
  });
};

export const fetchItems = ({ commit, dispatch, state }) => {
  commit(types.PRE_HTTP_REQUEST);
  api.fetchItems(state.apiEntryPoint, state.nextPage).then((data) => {
    dispatch('receiveItems', data);
  }).catch(() => {
    commit(types.FETCHED_ADS_FAIL);
  });
};

export const receiveItems = ({ commit }, data) => {
  commit(types.FETCHED_ADS_SUCCESS, data);
};

export const vote = ({ commit }, data) => {
  commit(types.BTN_LOADING);
  api.vote(data.choice).then((response) => {
    commit(types.VOTE, { response, data });
  }).catch((err) => {
    throw new Error(err);
  });
};

export const saveQuestion = ({ commit, state }, data) => {
  commit(types.BTN_LOADING);
  api.saveQuestion(state.apiEntryPoint, data).then((response) => {
    commit(types.QUESTION_SAVED, response);
  }).catch((err) => {
    throw new Error(err);
  });
};

