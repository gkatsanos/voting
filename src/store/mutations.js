import * as types from './mutation-types';
/* eslint-disable */
export default {
  [types.SET_URL](state, payload) {
    state.apiEntryPoint = payload;
  },
  [types.PRE_HTTP_REQUEST](state) {
    state.loadingState = true;
  },
  [types.BTN_LOADING](state) {
    state.btnLoadingState = true;
  },
  [types.FETCHED_ADS_SUCCESS](state, payload) {
    state.loadingState = false;
    state.questions = state.questions.concat(...payload);
    state.hasLoaded = true;
    state.nextPage += 1;
  },
  [types.FETCHED_ADS_FAIL](state) {
    state.loadingState = false;
    state.hasLoaded = true;
    state.nextPage = 1;
    state.maxPagesReached = true;
  },
  [types.VOTE](state, payload) {
    const myIndex = state.questions
      .findIndex(element => element.url === payload.data.question.url);
    const myChoiceIndex = state.questions[myIndex].choices
      .findIndex(element => element.url === payload.response.url);
    Object.assign(state.questions[myIndex].choices[myChoiceIndex], payload.response);
    state.btnLoadingState = false;
    state.questions[myIndex].voted = true;
  },
  [types.QUESTION_SAVED](state) {
    state.btnLoadingState = false;
    state.questionSaved = true;
    state.hasLoaded = false;
  },
  [types.RESET_QUESTION_FLAG](state) {
    state.questionSaved = false;
  },
};
