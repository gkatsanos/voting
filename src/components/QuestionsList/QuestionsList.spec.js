store.dispatch = jest.fn();
import Vue from 'vue';
import { mount } from 'vue-test-utils';
import QuestionsList from './QuestionsList';
import store from '@/store';

Vue.prototype.$vuetify = {
  load: (fn) => fn(),
  breakpoint: {}
};

describe('QuestionsList.spec.js', () => {
  let cmp;
  let template;
  
  beforeEach(() => {
    cmp = mount(QuestionsList, {
      store,
    });
    template = cmp.html();
  });
  
  it('has the expected html structure', () => {
    expect(template).toMatchSnapshot()
  });
  
  it('should emit 2 dispatch events', () => {
    expect(store.dispatch.mock.calls.length).toBe(2);
  });
  
});
