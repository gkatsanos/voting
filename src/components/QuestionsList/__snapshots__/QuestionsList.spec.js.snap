// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionsList.spec.js has the expected html structure 1`] = `
"<div><h4>Questions</h4> <v-layout row=\\"\\" wrap=\\"\\"></v-layout> <v-layout column=\\"\\" align-center=\\"\\"><v-btn large=\\"\\" color=\\"primary\\">
      Load More Questions</v-btn></v-layout></div>"
`;
